ttnn/cpp/ttnn/operations/ccl/all_gather_matmul/device/all_gather_matmul_op.cpp

```diff
@@ -0,0 +1,167 @@
+
+
+    // Matmul output tensor
+    ttnn::Tensor matmul_output_tensor = this->matmul_struct.create_output_tensors({input_tensors.at(1), input_tensors.at(2)}).at(0);
+
+    return {all_gather_output_tensor, matmul_output_tensor, datacopy_output_tensor};
+}
+
+operation::ProgramWithCallbacks AllGatherMatmul::create_program(const std::vector<Tensor> & input_tensors, const std::vector<std::optional<const ttnn::Tensor>>& optional_input_tensors, std::vector<Tensor> &output_tensors) const {
+
+    // Return the AllGatherMatmul program with callbacks
+    return all_gather_matmul_multi_core_with_workers(input_tensors.at(0), output_tensors.at(0), output_tensors.at(2), this->all_gather_struct.dim, this->all_gather_struct.num_links, this->all_gather_struct.ring_size, this->all_gather_struct.ring_index, this->all_gather_struct.receiver_device_id, this->all_gather_struct.sender_device_id, this->all_gather_struct.topology, this->all_gather_core_grid_offset);
+}
+
+namespace operations {
+namespace ccl {
+
+std::vector <ttnn::Tensor> all_gather_matmul(
+    const ttnn::Tensor& input_tensor,
+    const ttnn::Tensor& weight_tensor,
+    const uint32_t dim,
+    const CoreCoord all_gather_core_grid_offset,
+    const uint32_t num_links,
+    const std::optional<MemoryConfig>& memory_config,
+    const bool transpose_a,
+    const bool transpose_b,
+    const std::optional<const DataType> dtype,
+    const std::optional<const tt::operations::primary::MatmulProgramConfig> program_config,
+    const std::optional<const std::string>& activation,
+    const std::optional<const DeviceComputeKernelConfig> compute_kernel_config,
+    const std::optional<const ttnn::CoreGrid> core_grid
+) {
+
+    TT_FATAL(std::getenv("TT_METAL_SLOW_DISPATCH_MODE") == nullptr, "This op is only supported for Fast Dispatch");
+
+    auto devices = input_tensor.get_workers();
+    std::vector<Tensor> output_tensors = {ttnn::Tensor(operation::get_workers_for_op_output({input_tensor, weight_tensor})),
+                                            ttnn::Tensor(operation::get_workers_for_op_output({input_tensor, weight_tensor})),
+                                            ttnn::Tensor(operation::get_workers_for_op_output({input_tensor, weight_tensor}))};
+    std::vector<std::optional<const ttnn::Tensor>> optional_input_tensors = {std::nullopt};
+
+
+    operation::launch_op(
+        [dim, all_gather_core_grid_offset, num_links, memory_config, transpose_a, transpose_b, dtype, program_config, activation, compute_kernel_config, core_grid, devices](
+            const std::vector<Tensor>& input_tensors,
+            const std::vector<std::optional<const ttnn::Tensor>>& optional_input_tensors,
+            const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
+
+            const auto& input_tensor = input_tensors.at(0);
```

Comment:
This is from the all-gather launch-op, correct? If so, we should pull that out and reuse here to avoid potential of bugs/improvements being fixed in one place but not in another.